coroutines/src/commonTest/kotlin/DelayStrategyTest.kt

```diff
@@ -0,0 +1,102 @@
+    }
+
+    @Test
+    fun defaultsWithMinimumDelayMillis_numberBelowMinimum() {
+        assertEquals(1_000L, getExponentialBackoffMillis(baseTimeMillis = 10, minimumDelayMillis = 1_000L, iteration = 0))
+    }
+
+    @Test
+    fun firstIteration_hasDelayEqualToBaseTimeMillis() {
+        assertEquals(100, getExponentialBackoffMillis(baseTimeMillis = 100, iteration = 0))
+    }
+
+    @Test
+    fun multipleRetries_increaseExponentially() {
+        val baseTimeMillis = 100L
+        var delay = baseTimeMillis
+        val iterations = 0..10
+        iterations.forEach { iteration ->
+            assertEquals(delay, getExponentialBackoffMillis(baseTimeMillis = 100L, iteration = iteration), "Backoff retry #$iteration")
+            delay *= 2
+        }
+    }
+
+    @Test
+    @OptIn(ExperimentalTime::class)
+    fun dynamicDelayStrategy_switchesBetweenStrategies_usingTrigger() = runTest {
+        val trigger = MutableStateFlow(true)
+        val outputFlow: MutableStateFlow<Triple<Char, Int, Long>?> = MutableStateFlow(null)
+        // TestDelay that will push an 'A' test value to the output and suspend indefinitely.
+        val testDelayA = TestDelay { (iteration, elapsedMillis) ->
+            outputFlow.value = Triple('A', iteration, elapsedMillis)
+        }
+        // TestDelay that will push a 'B' test value to the output and suspend indefinitely.
+        val testDelayB = TestDelay { (iteration, elapsedMillis) ->
+            outputFlow.value = Triple('B', iteration, elapsedMillis)
+        }
+        // Switch between testDelayA and testDelayB, based on the trigger.
+        val selector: (Boolean) -> DelayStrategy = { if (it) testDelayA else testDelayB }
+
+        // TimeSource that will increment by a second every time ElapsedNow() is called.
+        val timeSource = object : TimeSource {
+            private var fakedTime = 1
+
+            override fun markNow() = object : TimeMark() {
+                override fun elapsedNow() = fakedTime++.toDuration(DurationUnit.SECONDS)
+            }
+        }
+
+        val delayJob = launch {
+            val dynamic = Dynamic(
+                trigger,
+                selector,
+                timeSource
+            )
+            dynamic.await(0, 0)
+        }
+
+        withTimeout(10_000) {
+            delay(100L)
+            assertEquals(Triple('A', 0, 1_000L), outputFlow.value)
+            trigger.value = false
+            delay(100L)
+            assertEquals(Triple('B', 0, 2_000L), outputFlow.value)
+            delayJob.cancelAndJoin()
+        }
+    }
+
+    private class TestDelay(
+        private val onAwait: (Pair<Int, Long>) -> Unit
+    ) : DelayStrategy {
+        override suspend fun await(iteration: Int, elapsedMillis: Long) {
+            onAwait(iteration to elapsedMillis)
+            awaitCancellation()
+        }
+    }
+}
```

Comment:
```suggestion
@OptIn(ExperimentalTime::class)
private class TimeMachine : TimeSource {

    private val elapsedMillis = MutableStateFlow(0L)

    fun advanceBy(milliseconds: Long) { elapsedMillis.update { value -> value + milliseconds } }
    suspend fun delayUntil(milliseconds: Long) { elapsedMillis.first { it >= milliseconds } }

    private class TimeMachineMark(private val parent: TimeMachine) : TimeMark() {
        override fun elapsedNow(): Duration = parent.elapsedMillis.value.toDuration(MILLISECONDS)
    }

    override fun markNow(): TimeMark = TimeMachineMark(this)
}

private class Invocation {
    data class Await(val iteration: Int, val elapsedMilliseconds: Long)
}

private class TimeMachineDelayStrategy(
    private val timeMachine: TimeMachine,
    private val delayUntilMilliseconds: Long,
) : DelayStrategy {

    val invocation = Channel<Invocation.Await>()

    override suspend fun await(iteration: Int, elapsedMillis: Long) {
        invocation.send(Invocation.Await(iteration, elapsedMillis))
        timeMachine.delayUntil(milliseconds = delayUntilMilliseconds)
    }
}
```

Reply:
Yeah I was playing around with it quite a bit yesterday, and a `Channel` totally makes sense here. I like it!